data/impl/RecordValidator.js

```diff
@@ -52,62 +52,104 @@ export class RecordValidator extends HoistBase {
     /** @return {number} - count of all validation errors for the record. */
     @computed
     get errorCount() {
-        return sumBy(this._validators, 'errorCount');
+        if (!this._fieldErrors) return 0;
+        return flatten(values(this._fieldErrors)).length;
     }
 
     /** @return {boolean} - true if any fields are currently recomputing their validation state. */
     @computed
     get isPending() {
-        return some(this._validators, it => it.isPending);
+        return this._validationTask.isPending;
     }
 
-    _validators = [];
+    @observable.ref _fieldErrors = {};
+    _validationTask = TaskObserver.trackLast();
+    _validationRunId = 0;
 
     /**
      * @param {Object} c - RecordValidator configuration.
      * @param {StoreRecord} c.record - record to validate
      */
     constructor({record}) {
-        super();
-        makeObservable(this);
         this.record = record;
-
-        const {fields} = this.record.store;
-        this._validators = fields.map(field => new RecordFieldValidator({record, field}));
+        makeObservable(this);
     }
 
     /**
      * Recompute validations for the record and return true if valid.
      * @returns {Promise<boolean>}
      */
     async validateAsync() {
-        const promises = map(this._validators, v => v.validateAsync());
-        await Promise.all(promises);
+        const runId = ++this._validationRunId,
+            fieldErrors = {},
+            {record} = this,
+            fieldsToValidate = record.store.fields.filter(it => !isEmpty(it.rules));
+
+        runInAction(() => this._fieldErrors = null);
+
+        const promises = fieldsToValidate.map(field => {
+            fieldErrors[field.name] = [];
+
+            const rulePromises = field.rules.map(async (rule) => {
+                const result = await this.evaluateRuleAsync(record, field, rule);
+                fieldErrors[field.name].push(result);
+            });
+
+            return Promise.all(rulePromises);
+        });
+
+        await Promise.all(promises).linkTo(this._validationTask);
+
+        runInAction(() => {
+            // TODO: Is this the correct place for this?
+            if (runId !== this._validationRunId) return;
+
+            this._fieldErrors = mapValues(fieldErrors, it => compact(flatten(it)));
+        });
+
         return this.isValid;
     }
 
     /** @return {ValidationState} - the current validation state for the record. */
     getValidationState() {
         const VS = ValidationState,
-            states = map(this._validators, v => v.validationState);
-        if (states.includes(VS.NotValid)) return VS.NotValid;
-        if (states.includes(VS.Unknown)) return VS.Unknown;
+            {_fieldErrors} = this;
+
+        // While executing any rules we are in an unknown validation state
+        if (_fieldErrors === null) return VS.Unknown;
```

Comment:
Isn't this going to cause a fully valid record to "flap" briefly to unknown, just when revalidating?  I think the existing setup was designed to avoid that intermediate state.